packages/flutter/lib/src/rendering/paragraph.dart

```diff
@@ -1616,15 +1620,32 @@ class _SelectableFragment with Selectable, ChangeNotifier implements TextLayoutM
     return result;
   }
 
-  TextPosition _getNextPosition(TextBoundary boundary, TextPosition position, bool forward) {
-    if (forward) {
-      return _clampTextPosition(
-        (PushTextPosition.forward + boundary).getTrailingTextBoundaryAt(position)
-      );
+  TextPosition _beyondTextBoundary(TextPosition extent, bool forward, TextBoundary textBoundary) {
+    final int newOffset = forward
+      ? textBoundary.getTrailingTextBoundaryAt(extent.offset) ?? range.end
+      : textBoundary.getLeadingTextBoundaryAt(extent.offset - 1) ?? range.start;
+    return TextPosition(offset: newOffset);
+  }
+
+  TextPosition _toTextBoundary(TextPosition extent, bool forward, TextBoundary textBoundary) {
```

Comment:
nit: we may need a better name or some doc string to document what this method is actually doing.

It looks like it converts the input from caret position to character position before passing into textboundary?

Reply:
also should the textposition to be store in character position in RenderParagraph instead so that we don't need to do this conversion?

Reply:
I mostly just copeid these 2 from editable text. Added comments and renamed the 2 methods.

Reply:
> also should the textposition to be store in character position in RenderParagraph instead so that we don't need to do this conversion?

I'll probably do that in a separate PR.